### PR TITLE
[WIP] Add context and timeouts

### PR DESCRIPTION
**Pull request by karlkfi**
- Add timeout flag to apply, destory, and preview cmds
- Default to no timeout
- Add context to TaskContext
- Pass context through to all functions that use it
- Rename prune-timeout flag to delete-timeout and add actual impl
- Rename PruneOptions to Pruner (there's already prune.Options)
- Rename PropagationPolicy to DeletionPropagationPolicy consistently
- Wrap func calls and defs for readability
- Remove Prune bool from solver.Options to simplify
  AppendPruneWaitTasks to always prune.
  Change Applier to only call AppendPruneWaitTasks if NoPrune.
  Change Destroyer to always call AppendPruneWaitTasks.
- Replace task.ClearTimeout with context.Cancel
- Add task.OnStatusEvent to fix WaitTask leaky abstraction.
  baseRunner was calling private funcs from WaitTask.
- Move ResourceStatusCollector into TaskContext. It doesn't need to
  persist between runs. Moving it simplifies baseRunner.
  This required making some of the collector funcs public.
- Remove Identifier from collector.ResourceStatus. Redundant with id.
- Add collector.Get/Put funcs for encapsulation and thread-safety
- Add locking to ResourceStatusCollector because a test was complaining
  about a race condition when updating status from another goroutine,
  like the poller does.
- Add logging to taskrunner tests for easier debugging concurrency
- Update task tests for cancel and timeout
- Rename options vars to "opts" for consistency
- Rewrite WaitTask context/timeout handling.
  TaskChannel is now only written to from one place, gated by an
  internal error channel and sync.Once.
  WaitTask now distinguishes between wait task timeout and parent
  context timeout.
- Update task tests to reflect new behavior with context cancel/timeout

**Note:** This PR is probably too big. I started cleaning things up and kept finding things to fix. Let me know what I should leave out or punt to another PR.

### CODE REVIEW

**Review by seans3 — CHANGES_REQUESTED** [review]
This is a massive PR, and it needs to be broken up in to safer, more digestible chunks. Renaming variables or functions should almost always be done in separate PR's to reduce risk and complexity. Smaller PR's are faster to merge **and** safer (as long as they are also well-tested).

On another note, do not worry about substantive changes to any code under the `cmd` hierarchy, since these commands are not exported; they are only used for testing and demonstration purposes.

**Comment by karlkfi** [comment]
Sure. I'll chop this into smaller PRs. Just wanted a place to stash a bunch of the tech debt fixes that I came up with while writing the apply-time-mutation PR. Was hoping to get some feedback on the stuff being done, but if it's too much, I'll pull stuff out, after I have time to figure out what's actually blocking the apply-time-mutation PR.

I think the ResourceCache and ResourceStatusCollector probably need to get merged. 

But of course, the e2e tests here are failing, so I must have broken something. I couldn't figure out how to even debug the failures on this branch, so I spent some time improving the tests and testutils to make it easier: https://github.com/kubernetes-sigs/cli-utils/pull/405

Hopefully once that's merged I can come back to this and figure out what went wrong.

**Comment by karlkfi** [comment]
/hold

This PR needs a lot of work. It's just here for staging until I have time to break it up.

**Comment by karlkfi** [comment]
Global timeout option extracted to https://github.com/kubernetes-sigs/cli-utils/pull/439

Still TODO:
- Default to no timeout (wait until cancelled)
- Add context to TaskContext & handle context cancellation in all task impls (instead of just WaitTask and between tasks)
- Rewrite WaitTask to use context cancellation
- Rename prune-timeout & delete-timeout to elimination-timeout ("elimination" parallels "reconciliation")
- Rename PruneOptions to Pruner (there's already prune.Options)
- Rename PropagationPolicy to DeletionPropagationPolicy consistently
- Clean up handling of optional Pruning
- Add Task.OnStatusEvent handler to fix WaitTask leaky abstraction.
- Remove Identifier from collector.ResourceStatus. Redundant with id.
- Rename options vars to "opts" for consistency

**Comment by karlkfi** [comment]
This is obsolete. The improvements have been made in other PRs and logged as backlog issues.

**Pull request closed without merging**
